discovery/src/main/java/com/alibaba/fescar/discovery/registry/consul/ConsulRegistryServiceImpl.java

```diff
@@ -0,0 +1,327 @@
+                }
+            }
+        }
+        return instance;
+    }
+
+    @Override
+    public void register(InetSocketAddress address) throws Exception {
+        NetUtil.validAddress(address);
+        getConsulClient().agentServiceRegister(createService(address));
+    }
+
+    @Override
+    public void unregister(InetSocketAddress address) throws Exception {
+        NetUtil.validAddress(address);
+        getConsulClient().agentServiceDeregister(createServiceId(address));
+    }
+
+    @Override
+    public void subscribe(String cluster, ConsulListener listener) throws Exception {
+        //1.add listener to subscribe list
+        listenerMap.putIfAbsent(cluster, new ArrayList<>());
+        listenerMap.get(cluster).add(listener);
+        //2.get healthy services
+        Response<List<HealthService>> response = getHealthyServices(cluster, -1, DEFAULT_WATCH_TIMEOUT);
+        //3.get current consul index.
+        Long index = response.getConsulIndex();
+        ConsulNotifier notifier = notifiers.computeIfAbsent(cluster, k -> new ConsulNotifier(cluster, index));
+        //4.run notifier
+        notifierExecutor.submit(notifier);
+    }
+
+    @Override
+    public void unsubscribe(String cluster, ConsulListener listener) throws Exception {
+        //1.remove notifier for the cluster
+        ConsulNotifier notifier = notifiers.remove(cluster);
+        //2.stop the notifier
+        notifier.stop();
+    }
+
+    @Override
+    public List<InetSocketAddress> lookup(String key) throws Exception {
+        final String cluster = getServiceGroup(key);
+        if (null == cluster) {
+            return null;
+        }
+        if (!listenerMap.containsKey(cluster)) {
+            //1.refresh cluster
+            refreshCluster(cluster);
+            //2. subscribe
+            subscribe(cluster, services -> {
+                refreshCluster(cluster, services);
+            });
+        }
+        return clusterAddressMap.get(cluster);
+    }
+
+    /**
+     * get consul client
+     *
+     * @return client
+     */
+    private ConsulClient getConsulClient() {
+        if (null == client) {
+            synchronized (ConsulRegistryServiceImpl.class) {
+                if (null == client) {
+                    client = new ConsulClient(FILE_CONFIG.getConfig(FILE_CONFIG_KEY_PREFIX + SERVER_ADDR_KEY));
+                }
+            }
+        }
+        return client;
+    }
+
+    /**
+     * get cluster name
+     *
+     * @return
+     */
+    private String getClusterName() {
+        String clusterConfigName = FILE_ROOT_REGISTRY + FILE_CONFIG_SPLIT_CHAR + REGISTRY_TYPE + FILE_CONFIG_SPLIT_CHAR
+            + REGISTRY_CLUSTER;
+        return FILE_CONFIG.getConfig(clusterConfigName, DEFAULT_CLUSTER_NAME);
+    }
+
+    /**
+     * create serviceId
+     *
+     * @param address
+     * @return serviceId
+     */
+    private String createServiceId(InetSocketAddress address) {
```

Comment:
How to ensure that multiple processes on a machine register the same cluster？

Reply:
done.will generate unique serviceId according to the host & port